### DISCUSSION TITLE
Worse performance vs V4?

### DISCUSSION THREAD

**Discussion by knubie** [ReactTooltip/react-tooltip#1006] [Q&A] [answered]
I'm migrating my app from v4 to v5 and I'm noticing some flickering / performance issues. For comparison:

V4:
https://user-images.githubusercontent.com/431251/230770121-5a8388e4-66ee-4226-87a1-7570a06ab577.mp4

V5:
https://user-images.githubusercontent.com/431251/230770135-71192e60-c891-401d-b21c-646b2c4124cf.mp4

Notice that in V5 the tooltip seems to stay on the previous anchor for a few milliseconds before moving to the new anchor, then disappears and reappears.

**Comment by gabrieljablonski** [comment]
This is most likely an issue with implementation on your end.

Are you able to share some samples from your project so we can give you some suggestions on how to improve this?

**Comment by knubie** [comment]
Sure. I haven't changed anything with the implementation other than updating the data-attributes on the anchors and the props on the `Tooltip` component. The only other difference would be that I'm not using the default CSS any more.

```js
<SidebarItem
  data-tooltip-id="sidebar"
  data-tooltip-content="Simple test"
  data-tooltip-place="right"
>
  Get Started
</SidebarItem>
// ...
<Tooltip id="sidebar" class="tooltip" />
```

It's also worth pointing out that the tooltip does not hide immediately on mouseout, there is a small delay, even when `delayHide` is set to `0`. If you have any clues as to what could be happening that would help me to debug I would much appreciate it.

**Reply by gabrieljablonski** [reply]
This probably won't change anything, but please also make sure you're running `react-tooltip@5.10.5`.

At first glance I can't see anything obvious that could be causing this problem, but here are a few ideas.

1. Try removing any custom CSS you have set on your application for `.tooltip` and `.react-tooltip` (this class is set internally)
2. Try moving the `<Tooltip />` component all the way up to the root of the DOM and see if that makes a difference

Aside from these, it's hard to tell what might be wrong without seeing your project. If it's a private project and you're unable to share it, you'd have to provide a sample project demonstrating the issue.

**Reply by knubie** [reply]
Thanks for the suggestions @gabrieljablonski . I put together this simple repro repo that illustrates the problem. It was set up using `create-react-app`.

https://github.com/knubie/react-tooltip-demo

Let me know if you are able to reproduce the issues in the demo video.

**Reply by gabrieljablonski** [reply, marked as answer]
> The only other difference would be that I'm not using the default CSS any more.

This should've been enough to diagnose it, my bad. As you can see on the [README file](https://github.com/ReactTooltip/react-tooltip#using-npm-package) (also on the docs), importing the default CSS file is NOT optional.

```jsx
import 'react-tooltip/dist/react-tooltip.css'
```

You can still use your `.tooltip` class to style the tooltip as you want, but (as you've just experienced) the default styling is NOT just cosmetic.

**Reply by knubie** [reply]
Aha, gotcha. Thanks for the help @gabrieljablonski . I think I was under the impression that the css was optional after reading this migration guide: https://react-tooltip.com/docs/upgrade-guide/changelog-v4-v5

> - CSS import is now optional, so you can modify and/or add any styling to your floating tooltip element

**Reply by gabrieljablonski** [reply]
Yeah that's our bad, I've already fixed that on a PR we'll be merging soon.
Thanks for reporting.

**Reply by knubie** [reply]
So, follow up question, is it possible to configure `react-tooltip` without the fade-in / delay?

**Reply by gabrieljablonski** [reply]
Have a look at the [internal styling](https://react-tooltip.com/docs/examples/styling#explanation) we use.

You should be able to override most stuff from `.tooltip` without breaking functionality, so `transform: none` should work on this case.

Just remember the class exposed externally is `.react-tooltip` (or use the `className` prop).

**Reply by knubie** [reply]
Thanks for the suggestion. I tried removing the `transition` property (I think that's what you meant) and setting `delayShow` and `delayHide` to `0` but there is still a noticeable delay on both.

I tracked the issue down to the debounced show and hide functions here: https://github.com/ReactTooltip/react-tooltip/blob/01c4a6fa1729b860639077d22dba8c47e1db808f/src/components/Tooltip/Tooltip.tsx#L170

The problem is that the functions are called on the trailing edge of the delay instead of the leading edge. Passing `true` as the `immediate` argument to `debounce` doesn't work either because the function itself does not appear to be implemented correctly.

**Reply by gabrieljablonski** [reply]
The debounce function does seem to be poorly implemented.

What are you calling noticeable? The extra 50 ms delay definitely isn't great, and we'll surely have a look at it, but is it that what you're referring to as noticeable, or are you seeing a larger delay?

If you don't mind, please open a new issue so we can track this more easily.

**Reply by knubie** [reply]
When I record the interaction and play frame by frame I count a 4\~5 frame (66.68\~83.35ms) delay.

https://user-images.githubusercontent.com/431251/230896643-e99c77cd-a01d-4251-970b-e3138f580ec4.mp4

----

When I remove the `debounce` call the delay drops to 1\~2 frames (16.67\~33.34ms) (not perfect, but acceptable).

https://user-images.githubusercontent.com/431251/230896734-7d1a0499-063a-4c94-bc8c-a6904a623ffb.mp4

I'm not sure what accounts for the extra 1-2 frames of delay. Could just be react itself? Would need to do some more experimenting probably.

Whether it is "noticeable" is, of course, subjective, but for me it is noticeable. Especially when running through a list:


https://user-images.githubusercontent.com/431251/230898153-0525ec7c-07a6-45d5-9d14-d4a6370c7e46.mp4


https://user-images.githubusercontent.com/431251/230898159-e2fa0a2d-5879-42f0-a50d-ac0a7011f668.mp4

**Reply by knubie** [reply]
Created an issue: https://github.com/ReactTooltip/react-tooltip/issues/1008
